ast_parser.py

```diff
@@ -0,0 +1,55 @@
+import json
+from ast_result import ASTResult
+
+# *Goal*
+# Parse an AST to return all the declared variables in the following format
+# {int}{myInt}
+# {string}{myInt}
+# {Foo}{myFooObject}
+
+
+class ASTParser:
+    def __init__(self):
+        print("ASTParser created")
+
+    # Load JSON file into dictionary and parse
+    def parse(self, file):
+        y = json.load(file)
+        ast_res = ASTResult()
+        root = y['Root']
+        varNodes = []
+        traverseSearch(root, 'VariableDeclaration', varNodes)  # Returns a list of nodes from the AST that are variables
+        ast_res.variableDeclarations = nodeToVar(varNodes)  # Parses each variable node to a variable tuple and returns a list of tuples
+        return ast_res
+
+# Converts each variable node to a variable tuple
+def nodeToVar(varNodes):
+    newList = []
+    for node in varNodes:
+        isArray = []
+        traverseSearch(node, 'ArrayCreationExpression', isArray)
+        varType = findVal(node, 'PredefinedType')
+        varName = findVal(node, 'VariableDeclarator')
+        if isArray:
+            varType += "[]"
+        newList.append((varType, varName))
+    return newList
+
+# Recursive traversal of AST to find a node with Type == lookfor
+# Appends all nodes that match to resultList which is maintained because python is pass by reference
```

Comment:
Is it because Python is "Pass-by-object-reference"? 
I have written "Pass-by-reference" because "Pass-by-reference" and "Pass-by-value" are the more common terms used.

Reply:
Correct, whereas most languages pass objects by reference, in Python objects are passed by value of reference. It usually does what you expect, but can sometimes bite you...